File: NotificationListener/app/src/main/java/com/example/notificationlistener/StatusBarParcelable.java
package com.example.notificationlistener;

import android.os.Parcel;
import android.os.Parcelable;
import android.service.notification.StatusBarNotification;
import android.util.Log;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

public class StatusBarParcelable implements Parcelable {
    private int id;
    private String appNotifier;
    private int tailleList;
    private ArrayList<String> key = new ArrayList<String>();
    public static String TAG = StatusBarParcelable.class.getSimpleName();

    public int getId() {
        return id;
    }

    public String getAppNotifier() {
        return appNotifier;
    }

    public int getTailleList() {
        return tailleList;
    }

    public ArrayList<String> getKey() {
        return key;
    }

    public HashMap<String, Object> getStatusBarNotificationInfo() {
        return statusBarNotificationInfo;
    }

    private HashMap<String,Object> statusBarNotificationInfo = new HashMap<String,Object>();

    public StatusBarParcelable(StatusBarNotification sbn) {
        id = sbn.getId();
        tailleList=sbn.getNotification().extras.keySet().size();
        appNotifier = sbn.getPackageName();
        for (String keySbn : sbn.getNotification().extras.keySet()) {
            key.add(keySbn);

            statusBarNotificationInfo.put(keySbn, sbn.getNotification().extras.get(keySbn));
        }
        //Log.i(TAG, "ID:" + sbn.getId());
        //Log.i(TAG, "Posted by:" + sbn.getPackageName());


    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(id);
        dest.writeString(appNotifier);
        dest.writeInt(tailleList);

        Set<Map.Entry<String, Object>> entree = statusBarNotificationInfo.entrySet();

        for(Map.Entry<String,Object> e:entree){

                dest.writeString(e.getKey());
                dest.writeString(" "+e.getValue());

        }
        dest.writeString(appNotifier);
        //dest.writeList(key);
        //dest.writeMap(statusBarNotificationInfo);
    }

    public static Creator<StatusBarParcelable> CREATOR = new Creator<StatusBarParcelable>() {
        @Override
        public StatusBarParcelable createFromParcel(Parcel in) {
            return new StatusBarParcelable(in);
        }
        @Override
        public StatusBarParcelable[] newArray(int size) {
            return new StatusBarParcelable[size];
        }
    };

    private StatusBarParcelable(Parcel in) {
        id=in.readInt();
        appNotifier=in.readString();
        tailleList=in.readInt();

        Log.i(TAG,"posted by:"+appNotifier+" taille list"+tailleList);
        for (int i=0;i<tailleList;i++){
            String keysbn= in.readString();
            String value= in.readString();
            key.add(keysbn);
            statusBarNotificationInfo.put(keysbn,value);
                if (key != null) {
                    Log.i(TAG, keysbn + "=" + this.getStatusBarNotificationInfo().get(keysbn));
                }
            }
        }
        //key = in.readArrayList(String.class.getClassLoader());
        //statusBarNotificationInfo=in.readHashMap(String.class.getClassLoader());

}

File: NotificationListener/app/src/main/java/com/example/notificationlistener/StatusBarParcelable.kt
package com.example.notificationlistener

import android.os.Parcel
import android.os.Parcelable
import android.service.notification.StatusBarNotification
import android.util.Log

import java.util.ArrayList
import java.util.HashMap

class StatusBarParcelable : Parcelable {
    var id: Int = 0
        private set
    var appNotifier: String? = null
        private set
    var tailleList: Int = 0
        private set
    val key = ArrayList<String>()

    val statusBarNotificationInfo = HashMap<String, Any>()

    constructor(sbn: StatusBarNotification) {
        id = sbn.id
        tailleList = sbn.notification.extras.keySet().size
        appNotifier = sbn.packageName
        for (keySbn in sbn.notification.extras.keySet()) {
            key.add(keySbn)
            if(sbn.notification.extras.get(keySbn)!= null) {
                statusBarNotificationInfo[keySbn] = sbn.notification.extras.get(keySbn)
            }else{
                statusBarNotificationInfo[keySbn]= ""
            }
        }
        //Log.i(TAG, "ID:" + sbn.getId());
        //Log.i(TAG, "Posted by:" + sbn.getPackageName());


    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(id)
        dest.writeString(appNotifier)
        dest.writeInt(tailleList)

        val entree = statusBarNotificationInfo.entries

        for ((key1, value) in entree) {

            dest.writeString(key1)
            dest.writeString(" $value")

        }
        dest.writeString(appNotifier)
        //dest.writeList(key);
        //dest.writeMap(statusBarNotificationInfo);
    }

    private constructor(`in`: Parcel) {
        id = `in`.readInt()
        appNotifier = `in`.readString()
        tailleList = `in`.readInt()

        Log.i(TAG, "posted by:$appNotifier taille list$tailleList")
        for (i in 0 until tailleList) {
            val keysbn = `in`.readString()
            val value = `in`.readString()
            key.add(keysbn)
            statusBarNotificationInfo[keysbn] = value
            if (key != null) {
                Log.i(TAG, keysbn + "=" + this.statusBarNotificationInfo[keysbn])
            }
        }
    }

    companion object CREATOR: Parcelable.Creator<StatusBarParcelable> {
        override fun createFromParcel(source: Parcel): StatusBarParcelable {
            return StatusBarParcelable(source)
        }

        override fun newArray(size: Int): Array<StatusBarParcelable?> {
            return arrayOfNulls(size)
        }

    }
        var TAG = StatusBarParcelable::class.java.simpleName


    }
